### DISCUSSION TITLE
Confusion around purge/content configuration and missing styles in production

### DISCUSSION THREAD

**Discussion by guillaumecherel** [tailwindlabs/tailwindcss#7568] [Help] [answered]
I followed the instructions in [Install Tailwind CSS with Create React App](https://tailwindcss.com/docs/guides/create-react-app) which worked great for development, but ran into problems that I had a hard time to figure out when deploying in production. I finally found that the issue is related to purging the CSS. I'm putting my report here hoping it to be useful to others.

My app is fetching some Markdown file with mathjax syntax and displaying it to the user. To control how the math is rendered, I added some lines to a [base style](https://tailwindcss.com/docs/adding-base-styles). The math displayed properly when I started the development server with `npm start`, but didn't when building for production with `npm run build`. 

After many silly hours (ok, days) of head scratching, I started reading tailwinds doc in deep and learned about PurgeCSS. I realized that the problem came from purging the css classes in the tailwind configuration. The css classes that are generated from parsing the markdown and generating html are not present at build time, and thus purged.

The page [Install Tailwind CSS with Create React App](https://tailwindcss.com/docs/guides/create-react-app) instructs you to add those lines to the configuration:

```
  // tailwind.config.js
  module.exports = {
   purge: ['./src/**/*.{js,jsx,ts,tsx}', './public/index.html'],

   ...
  }
```

I did it without thinking much about it, and not being extensively versed in the frontend development universe, I wasn't smart enough to light up a mental warning by myself when I read those lines.
 
Perhaps it'd help others like me to explicitly warn the reader that any class that doesn't appear in the code at **compile time** will be purged from the final css.

**Comment by simonswiss** [comment, marked as answer]
Hey!

Sorry to hear about the time-consuming experience. This is definitely a friction area for some users.

Technically, Tailwind CSS since v3 does not use Purge CSS at all anymore: instead of removing unused styles for production, Tailwind only generates the styles you're using, with _just in time_ compilation.

The `purge` configuration somewhat still exists - i[t's now called `content`](https://tailwindcss.com/docs/content-configuration) and is necessary for Tailwind to know _where_ to look in a project to generate the CSS output.

If that helps, I am right now working on a series of "Setting up Tailwind CSS in an `X` project" videos for many popular frameworks. In each of those screencasts, I go into quite a bit of detail about the `content` configuration, which should help a lot with the confusion you're describing.

Instead of just closing this issue, I'm going to convert it to a GitHub discussion, as this might help other folks out there going through the same struggle.

Thanks for reporting this, and have a great day!

**Comment by reesz** [comment]
So I am getting to this over a year later with a similar issue. @simonswiss is there any update on it?

I'm using Tailwind to develop a WordPress theme which tries to include styles for different kind of WordPress plugins or even just npm plugins that are used throughout the theme.

Obviously not all classes are present at the time of the JIT compiler, so a lot of classes are missing, even ones that are explicitly created in my style.scss.

In my use case a simple way of marking classes as "non-purgable" while writing the SCSS could fix this, like:
```scss
@mark no-purge {
  .medium-zoom-overlay {
    z-index: 90;
  }
}
```

Or however the syntax may look like.
I'm also using a plugin function to generate WordPress compatible classes for all Tailwind colors like this:
```js
function( { addBase, theme } ) {
            function generateColorUtilities( colorObj, colorGroup = '' ) {
                const utilities = {};

                for ( const colorKey in colorObj ) {
                    const value = colorObj[colorKey];
                    const slug  = `${colorGroup}-${colorKey}`;

                    if ( typeof value === 'string' ) {
                        utilities[`.has${slug}-color`]            = { color: value };
                        utilities[`.has${slug}-background-color`] = { backgroundColor: value };
                    }
                    else {
                        const nestedUtilities = generateColorUtilities(value, slug);
                        Object.assign(utilities, nestedUtilities);
                    }
                }

                return utilities;
            }

            const colorUtilities = generateColorUtilities(theme('colors'));
            addBase(colorUtilities);
        }
```

which also get purged on JIT.

I think it's safe to assume, that if someone manually adds classes to an output, they should stay in?
Or add another layer similar to base/utilities/components that could be used to store non-purgable classes?

**Reply by wongjn** [reply]
For class rules you define directly in CSS, you could define them outside a `@layer` (if you are not using any variants with them) and they will be preserved in the compiled CSS:

```css
.medium-zoom-overlay {
  z-index: 90;
}
```

Otherwise, and for classes generated by plugins, you can [safelist classes](https://tailwindcss.com/docs/content-configuration#safelisting-classes) you'd like to see in the compiled CSS.

**Reply by reesz** [reply]
Ah perfect, I remembered Tailwind/Webpack giving me errors when I declared anything outside of a `@layer` before, may have been from older version, so that habit stuck to place all SCSS imports into their layers.

Will be looking into these options, thanks!
